### ISSUE TITLE
Better Implementation in Activation Grids Notebook

### ISSUE THREAD

**Issue by ShiRuiUT** [tensorflow/lucid#139]
Hi, thanks for your previous helps. There is a little trouble in index selecting of Better Implementation in Activation Grids.

The method to decrease computing consumption is to compute 64/196 batches by the code:
`rand_inds = tf.random_uniform([batch_size], 0, N, dtype=tf.int32)`.
But when I change the rand_inds to `rand_inds = tf.range(0, 64, 1)`, I found indexes I never selected are updated somewhat. I thought they should be random noise.

Reason I did this test is because running times of naive implement and better implement are not much different. Both implements took about 350s in my computer GPU 1080. (Theoretically, two running times should have several times gap I think)
Could you help me to make some analysis about this trouble? Thank you very much.

indexes: 0-63
![stackedactgridsbetter2](https://user-images.githubusercontent.com/19572042/52462006-b4446680-2bb4-11e9-912d-04a322779a9e.jpeg)

random 64 indexes
![stackedactgridsbetter](https://user-images.githubusercontent.com/19572042/52462032-c7efcd00-2bb4-11e9-8b69-f86ff4654955.jpeg)

**Comment by ShiRuiUT** [comment]
Hello, sorry submit two things in one issue.
The thing is when I read your paper (https://distill.pub/2018/building-blocks/) about this tutorial, I found 
> For instance, GoogLeNet has multiple floppy ear detectors that appear to detect slightly different levels of droopiness, length, and surrounding context to the ears.
In fact, when I handle some more advanced network (ResNext, SE-ResNeXt), I also found there are many similarly functional convolutional kernels in lowest and highest layers like paper said.

Have you tried some methods to suppress appearance of these similar activation maps when training a new network? Or maybe it could improve performance. 

I tried to introduce correlation coefficient (SSIM) and wasserstein-1 distance to evaluate similarity among activation maps. But there is no idea to combine these methods into loss function for training in a fast enough way. I would very appreciate it if you would like share some thoughts or related papers about it.

**Comment by colah** [comment]
> The method to decrease computing consumption is to compute 64/196 batches by the code:
`rand_inds = tf.random_uniform([batch_size], 0, N, dtype=tf.int32)`.
But when I change the rand_inds to `rand_inds = tf.range(0, 64, 1)`, I found indexes I never selected are updated somewhat. I thought they should be random noise.

That's a nice way to test your understanding! You're absolutely right that that is a surprising result. The issue is a bit subtle.

If you do an activation grid visualization naively, it will have snouts and such facing different directions in each place. This isn't because they are representing different vectors, it's that the model mostly doesn't care at that point about orientation.

![image](https://user-images.githubusercontent.com/61658/52487259-17f97000-2b72-11e9-84fb-512640a81b55.png)

In order to fix this, we use a special parameterization. This general idea is discussed in the [differentiable parameterizations](https://distill.pub/2018/differentiable-parameterizations/) article, which you may enjoy reading. In particular, [aligned feature interpolation](https://distill.pub/2018/differentiable-parameterizations/#section-aligned-interpolation) is a kind of similar idea.

For this particular visualization, we use the following parametrization:

```py
    group_imgs_raw = param.fft_image([n_groups, W, W, 3])
    unique_imgs_raw = param.fft_image([N, W, W, 3])
    opt_imgs = param.to_valid_rgb(tf.stack([
            0.7*unique_imgs_raw[i] + 
            0.5*sum(groups[i, j] * group_imgs_raw[j] for j in range(n_groups))
            for i in range(N) ]),
            decorrelate=True)
```

Because `group_imgs_raw` are shared between all the images we optimize, updating some images will also cause others to change.

**Comment by colah** [comment]
> Have you tried some methods to suppress appearance of these similar activation maps when training a new network?

I think you are asking if we've tried to train a network while suppressing, say, it from having a floppy ear detector? I have not tried anything like that, but it sounds like an interesting experiment!

**Comment by ShiRuiUT** [comment]
Hi, thanks to your fast and warm response. 
First one:
It seems like tensorflow updating all the variables in Graph, so the related tensor that makes up `opt_imgs` will be updated for updating `opt_imgs`. That caused others to change, it sounds a bit counterintuitive but reasonable. 

When I initialized `group_imgs_raw` in advance, wanted results were generated.
![stackedactgridsbetter3](https://user-images.githubusercontent.com/19572042/52516497-ca632d00-2c6f-11e9-8846-22820c60af69.jpeg)

Second one:
Say, there are always multiple floppy ear detectors appeared. In other words, these detectors' outputs or activation maps will be similar (similar channel-wise matrix). There are many ways to evaluate similarity between matrix or vector. 

I think suppressing appearance of multiple similar detectors is approximately equal to decreasing similarity between activation maps. Combining the way of evaluating activation maps with loss function maybe improve performance. 

But I don't have further idea about this, e.g., how to solve the problem of a surge in computing, because there are 512 activation maps that need to compare similarities. I think of this problem, because matrix decomposition methods are used in your code for dimensionality reduction.

**Comment by ShiRuiUT** [comment]
There is a paper in CVPR2019 seems have begun to solve **conv-kernel functionally overlap or redundancy problem**. 
_RePr: Improved Training of Convolutional Filters. A Prakash, J Storer, D Florencio, C Zhang. CVPR2019._
This is a good idea to solve the problem.

Further, I think better kernel overlap metric should be derived from better understanding of conv-kernel functions. Better understanding -> better **metric** for kernel functionally overlap -> alter the training method by fusing with the **metric** (like this paper's metric and training method).
